### ISSUE TITLE
Guide for Discriminators (Nested and Non-Nested)

### ISSUE THREAD

**Issue by hasezoey** [typegoose/typegoose#565]
documentation on discriminators (nested and non-nested) are currently both lacking

Todo / Finished:

- [ ] [Nested Discriminators](https://typegoose.github.io/typegoose/docs/api/decorators/prop/#discriminators)
- [x] [Non-Nested Discriminators](https://typegoose.github.io/typegoose/docs/api/functions/get-discriminator-model-for-class/)

Pull Requests are welcome (it should be added under `/docs/guides/advanced`)

**Comment by meepeek** [comment]
---
id: non-nested-discriminators
title: 'Non-Nested Discriminators'
---

## Non-Nested Discriminators

### TLDR version
1. Put ```@modelOptions({ schemaOptions: { discriminationKey: 'type' } })``` on top of models's class (see example below)
2. Use getDiscriminatorModelForClass instead of getModelForClass

### Detail version
Consider a use case where you need multiple type of documents share the same database collection so you can query from the same place. However, you will want typegoose to distinct the object being queried in a different models. This is where Discriminators feature comes in.

Let's imagine, you are a veterinarian and you want to store data for your clients' pet. You want to store both Cat and Dog data in Animal collection so you don't need write a code to lookup for them separately.  Typegoose's model can then be represented as follow:
```ts
class Animal {
	@prop({ required:  true, unique: true })
	public: name!: string
}
```
You can then extended the Animal for Cat and Dog.
```ts
@modelOptions({ schemaOptions: { collection:  'animal' } })
class Dog extends Animal {
	@prop()
	cageNumber!: number
}
@modelOptions({ schemaOptions: { collection:  'animal' } })
class Cat extends Animal {
	@prop()
	nameTag!: string
}
```
Now you can create Dog and Cat separately with cat.create and dog.create. both of them will store in database collection named 'animal'

The problem arise when you want to lookup using animal.findOne() since you'll get the record, but there would be no cageNumber or nameTag property since class Animal does not contain one. You will get type error if you try to access them.
```ts
const cat = getModelForClass(Cat)
await cat.create({name: 'petty', nameTag: 'petty de cat'})
const dog = getModelForClass(Dog)
await dog.create({name: 'pochi', cageNumber: 5})
const animal = getModelFroClass(Animal)
console.log(await animal.findOne()) // {name: 'petty'} no nameTag property
```
Also, calling animal.create, you will not be able to pass cageName or nameTag value for the same reason. You need a way to tell typegoose to select the classes for you, based on the type. This is where Discriminators come in handy.
```ts
@modelOptions({
    schemaOptions: {
        discriminatorKey: 'type'
    }
})
class Animal {
	@prop({ required:  true, unique: true })
	public: name!: string

	@prop({ required:  true })
	public: type!: string
}
```
After that, you will use getDiscriminatorModelForClass instead of getModelForClass when you get the model. You will then be able to create and find both Cat and Dog from Animal model
```ts
const animal = getDiscriminatorModelForClass(Animal)
```

**Comment by meepeek** [comment]
This is an incomplete draft. I wrote based on my current understanding after talking with hasezoey I will need to test the use case in my project so I can fill a correct code at the end.

@hasezoey please review the content and let me know if you have any suggestion

**Comment by hasezoey** [comment]
@meepeek next time please do an PR or a gist file, and not an comment here for an draft

aside from that, a small review:
- from what i know, if you dont define "collection" (or define it just on the parent), and use it with `getDiscriminatorModelForClass`, it will by default share the same (and if you define it on the parent and use `getModelForClass`, it will also share it to extending classes if not overwriting)

**Comment by hasezoey** [comment]
@meepeek i made an guide #566 , do you think this is an ok version?

**Comment by meepeek** [comment]
I think it's very good. I have a better understanding when I read it.

**Comment by ThatOneAwkwardGuy** [comment]
@hasezoey do you also have any opinions on the best way to get the discriminant model when updating an entity?

e.g if I were to use animal.discriminants, the attributes of that object are the class names instead of the discriminatorId set by the TypegooseModule from 'nestjs-typegoose'

**Comment by hasezoey** [comment]
@ThatOneAwkwardGuy please open another issue / discussion, or write on discord, this issue here is for tracking discriminator guides (not help)

**Comment by ThatOneAwkwardGuy** [comment]
> @ThatOneAwkwardGuy please open another issue / discussion, or write on discord, this issue here is for tracking discriminator guides (not help)

gotcha, my bad :)

**Comment by hasezoey** [comment]
closing because:

- the guide for "Non-Nested Discriminators" was added in 27f879090aae9023519f16c5826f63e9b3deb6d4
- the guide for "Nested Discriminators" was added in 1fcb7d7a0ee76933e05bdd0d098a3d6c3db1b518
